File: SDKs/web3/src/IVXHiroSystems.ts
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

// ---------------------------------------------------------------------------
// Hiro Systems — Spin wheel, streaks, offerwall, friends (Web3 SDK)
// ---------------------------------------------------------------------------

import { Client, Session } from '@heroiclabs/nakama-js';

export interface IVXSpinWheelReward {
  rewardId: string;
  type: string;
  amount: number;
  metadata?: Record<string, unknown>;
}

export interface IVXStreakInfo {
  currentStreak: number;
  longestStreak: number;
  claimedToday: boolean;
  lastClaimDate: string;
  rewards: IVXSpinWheelReward[];
}

export interface IVXOfferwallItem {
  offerId: string;
  title: string;
  description: string;
  imageUrl: string;
  cost: number;
  currency: string;
  available: boolean;
}

export interface IVXFriendInfo {
  userId: string;
  username: string;
  displayName: string;
  avatarUrl: string;
  state: number;
}

export interface IVXHiroEventMap {
  wheelSpun: [reward: IVXSpinWheelReward];
  streakClaimed: [info: IVXStreakInfo];
  offerClaimed: [offerId: string];
  friendAdded: [userId: string];
  friendRemoved: [userId: string];
  error: [error: { code: number; message: string }];
}

type HiroEventHandler<K extends keyof IVXHiroEventMap> = (...args: IVXHiroEventMap[K]) => void;

/**
 * Hiro-powered game systems accessed via Nakama RPC.
 *
 * Wraps spin wheel, daily streaks, offerwall, and friend management
 * RPCs with typed interfaces for the Web3 SDK.
 */
export class IVXHiroSystems {
  private static _instance: IVXHiroSystems | null = null;

  private _client: Client | null = null;
  private _session: Session | null = null;
  private _enableDebugLogs = false;
  private _listeners = new Map<string, Set<Function>>();

  /** Return the shared singleton instance. */
  static getInstance(): IVXHiroSystems {
    if (!IVXHiroSystems._instance) {
      IVXHiroSystems._instance = new IVXHiroSystems();
    }
    return IVXHiroSystems._instance;
  }

  /** Reset the singleton (useful for testing). */
  static resetInstance(): void {
    IVXHiroSystems._instance = null;
  }

  private constructor() {}

  /**
   * Bind a Nakama client and session for RPC calls.
   * Typically called after IVXWeb3Manager authenticates.
   */
  configure(client: Client, session: Session, enableDebugLogs = false): void {
    this._client = client;
    this._session = session;
    this._enableDebugLogs = enableDebugLogs;
    this.log('Hiro systems configured');
  }

  // ---------------------------------------------------------------------------
  // Events
  // ---------------------------------------------------------------------------

  /** Subscribe to a Hiro event. */
  on<K extends keyof IVXHiroEventMap>(event: K, handler: HiroEventHandler<K>): void {
    if (!this._listeners.has(event)) {
      this._listeners.set(event, new Set());
    }
    this._listeners.get(event)!.add(handler);
  }

  /** Unsubscribe from a Hiro event. */
  off<K extends keyof IVXHiroEventMap>(event: K, handler: HiroEventHandler<K>): void {
    this._listeners.get(event)?.delete(handler);
  }

  private emit<K extends keyof IVXHiroEventMap>(event: K, ...args: IVXHiroEventMap[K]): void {
    this._listeners.get(event)?.forEach(fn => (fn as Function)(...args));
  }

  // ---------------------------------------------------------------------------
  // Spin Wheel
  // ---------------------------------------------------------------------------

  /** Spin a prize wheel and return the reward. */
  async spinWheel(wheelId = 'default'): Promise<IVXSpinWheelReward> {
    const result = await this.callRpc('hiro_spin_wheel', JSON.stringify({ wheel_id: wheelId }));
    const reward: IVXSpinWheelReward = {
      rewardId: result.reward_id ?? '',
      type: result.type ?? '',
      amount: Number(result.amount ?? 0),
      metadata: result.metadata as Record<string, unknown> | undefined,
    };
    this.log(`Wheel spun: ${reward.type} x${reward.amount}`);
    this.emit('wheelSpun', reward);
    return reward;
  }

  /** Fetch the wheel configuration (segments, probabilities). */
  async getWheelConfig(wheelId = 'default'): Promise<Record<string, unknown>> {
    return this.callRpc('hiro_spin_wheel_config', JSON.stringify({ wheel_id: wheelId }));
  }

  // ---------------------------------------------------------------------------
  // Daily Streaks
  // ---------------------------------------------------------------------------

  /** Get the current streak state for the authenticated user. */
  async getStreak(): Promise<IVXStreakInfo> {
    const result = await this.callRpc('hiro_streak_get', '{}');
    return this.parseStreakInfo(result);
  }

  /** Claim today's streak reward. */
  async claimStreak(): Promise<IVXStreakInfo> {
    const result = await this.callRpc('hiro_streak_claim', '{}');
    const info = this.parseStreakInfo(result);
    this.log(`Streak claimed — day ${info.currentStreak}`);
    this.emit('streakClaimed', info);
    return info;
  }

  // ---------------------------------------------------------------------------
  // Offerwall
  // ---------------------------------------------------------------------------

  /** List available offerwall items. */
  async getOffers(): Promise<IVXOfferwallItem[]> {
    const result = await this.callRpc('hiro_offerwall_list', '{}');
    const offers: IVXOfferwallItem[] = Array.isArray(result.offers)
      ? result.offers.map((o: Record<string, unknown>) => ({
          offerId: String(o.offer_id ?? ''),
          title: String(o.title ?? ''),
          description: String(o.description ?? ''),
          imageUrl: String(o.image_url ?? ''),
          cost: Number(o.cost ?? 0),
          currency: String(o.currency ?? ''),
          available: Boolean(o.available ?? true),
        }))
      : [];
    return offers;
  }

  /** Claim (purchase) an offerwall item. */
  async claimOffer(offerId: string): Promise<void> {
    await this.callRpc('hiro_offerwall_claim', JSON.stringify({ offer_id: offerId }));
    this.log(`Offer claimed: ${offerId}`);
    this.emit('offerClaimed', offerId);
  }

  // ---------------------------------------------------------------------------
  // Friends
  // ---------------------------------------------------------------------------

  /** List friends, optionally filtered by state (0=mutual, 1=invite sent, etc.). */
  async listFriends(state = 0): Promise<IVXFriendInfo[]> {
    const result = await this.callRpc('hiro_friends_list', JSON.stringify({ state }));
    const friends: IVXFriendInfo[] = Array.isArray(result.friends)
      ? result.friends.map((f: Record<string, unknown>) => ({
          userId: String(f.user_id ?? ''),
          username: String(f.username ?? ''),
          displayName: String(f.display_name ?? ''),
          avatarUrl: String(f.avatar_url ?? ''),
          state: Number(f.state ?? 0),
        }))
      : [];
    return friends;
  }

  /** Send a friend request. */
  async addFriend(userId: string): Promise<void> {
    await this.callRpc('hiro_friends_add', JSON.stringify({ user_id: userId }));
    this.log(`Friend added: ${userId}`);
    this.emit('friendAdded', userId);
  }

  /** Remove a friend. */
  async removeFriend(userId: string): Promise<void> {
    await this.callRpc('hiro_friends_remove', JSON.stringify({ user_id: userId }));
    this.log(`Friend removed: ${userId}`);
    this.emit('friendRemoved', userId);
  }

  /** Block a user. */
  async blockUser(userId: string): Promise<void> {
    await this.callRpc('hiro_friends_block', JSON.stringify({ user_id: userId }));
    this.log(`User blocked: ${userId}`);
  }

  // ---------------------------------------------------------------------------
  // Internal
  // ---------------------------------------------------------------------------

  private async callRpc(rpcId: string, payload = '{}'): Promise<Record<string, unknown>> {
    this.ensureConfigured();
    try {
      const result = await this._client!.rpc(this._session!, rpcId, payload);
      this.log(`RPC ${rpcId} response received`);
      return result.payload ? this.safeParseJson(result.payload) : {};
    } catch (e: unknown) {
      const error = this.toError(e);
      this.emit('error', error);
      throw error;
    }
  }

  private parseStreakInfo(data: Record<string, unknown>): IVXStreakInfo {
    return {
      currentStreak: Number(data.current_streak ?? 0),
      longestStreak: Number(data.longest_streak ?? 0),
      claimedToday: Boolean(data.claimed_today ?? false),
      lastClaimDate: String(data.last_claim_date ?? ''),
      rewards: Array.isArray(data.rewards)
        ? data.rewards.map((r: Record<string, unknown>) => ({
            rewardId: String(r.reward_id ?? ''),
            type: String(r.type ?? ''),
            amount: Number(r.amount ?? 0),
          }))
        : [],
    };
  }

  private ensureConfigured(): void {
    if (!this._client || !this._session) {
      const err = { code: -1, message: 'Hiro systems not configured. Call configure() first.' };
      this.emit('error', err);
      throw err;
    }
  }

  private safeParseJson(value: unknown): Record<string, unknown> {
    if (typeof value === 'object' && value !== null) return value as Record<string, unknown>;
    if (typeof value !== 'string' || value === '') return {};
    try { return JSON.parse(value); } catch { return {}; }
  }

  private toError(e: unknown): { code: number; message: string } {
    if (typeof e === 'object' && e !== null && 'code' in e && 'message' in e) {
      return e as { code: number; message: string };
    }
    if (e instanceof Error) return { code: -1, message: e.message };
    return { code: -1, message: String(e) };
  }

  private log(message: string): void {
    if (this._enableDebugLogs) {
      console.log(`[IntelliVerseX Web3:Hiro] ${message}`);
    }
  }
}

File: SDKs/cocos2dx/Classes/IntelliVerseX/IVXHiroSystems.cpp
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

#include "IntelliVerseX/IVXHiroSystems.h"
#include "cocos2d.h"
#include "json/rapidjson.h"
#include "json/document.h"
#include "json/writer.h"
#include "json/stringbuffer.h"

namespace IntelliVerseX {

IVXHiroSystems& IVXHiroSystems::getInstance() {
    static IVXHiroSystems instance;
    return instance;
}

// ---------------------------------------------------------------------------
// Spin Wheel
// ---------------------------------------------------------------------------

void IVXHiroSystems::spinWheel(const std::string& wheelId,
                                SpinWheelCallback onSuccess,
                                ErrorCallback onError) {
    std::string payload = "{\"wheel_id\":\"" + wheelId + "\"}";
    callHiroRpc("hiro_spin_wheel", payload,
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            SpinWheelReward reward;
            if (!doc.HasParseError() && doc.IsObject()) {
                reward.rewardId = doc.HasMember("reward_id") ? doc["reward_id"].GetString() : "";
                reward.type = doc.HasMember("type") ? doc["type"].GetString() : "";
                reward.amount = doc.HasMember("amount") ? doc["amount"].GetInt() : 0;
                reward.metadata = doc.HasMember("metadata") ? doc["metadata"].GetString() : "";
            }
            if (onSuccess) onSuccess(reward);
        }, onError);
}

void IVXHiroSystems::getWheelConfig(const std::string& wheelId,
                                     RpcCallback onSuccess,
                                     ErrorCallback onError) {
    std::string payload = "{\"wheel_id\":\"" + wheelId + "\"}";
    callHiroRpc("hiro_spin_wheel_config", payload, onSuccess, onError);
}

// ---------------------------------------------------------------------------
// Daily Streaks
// ---------------------------------------------------------------------------

void IVXHiroSystems::getStreak(StreakCallback onSuccess,
                                ErrorCallback onError) {
    callHiroRpc("hiro_streak_get", "{}",
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            StreakInfo info;
            if (!doc.HasParseError() && doc.IsObject()) {
                info.currentStreak = doc.HasMember("current_streak") ? doc["current_streak"].GetInt() : 0;
                info.longestStreak = doc.HasMember("longest_streak") ? doc["longest_streak"].GetInt() : 0;
                info.claimedToday = doc.HasMember("claimed_today") ? doc["claimed_today"].GetBool() : false;
                info.lastClaimDate = doc.HasMember("last_claim_date") ? doc["last_claim_date"].GetString() : "";
            }
            if (onSuccess) onSuccess(info);
        }, onError);
}

void IVXHiroSystems::claimStreak(StreakCallback onSuccess,
                                  ErrorCallback onError) {
    callHiroRpc("hiro_streak_claim", "{}",
        [this, onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            StreakInfo info;
            if (!doc.HasParseError() && doc.IsObject()) {
                info.currentStreak = doc.HasMember("current_streak") ? doc["current_streak"].GetInt() : 0;
                info.longestStreak = doc.HasMember("longest_streak") ? doc["longest_streak"].GetInt() : 0;
                info.claimedToday = true;
                info.lastClaimDate = doc.HasMember("last_claim_date") ? doc["last_claim_date"].GetString() : "";
            }
            log("Streak claimed — day " + std::to_string(info.currentStreak));
            if (onSuccess) onSuccess(info);
        }, onError);
}

// ---------------------------------------------------------------------------
// Offerwall
// ---------------------------------------------------------------------------

void IVXHiroSystems::getOffers(OfferwallCallback onSuccess,
                                ErrorCallback onError) {
    callHiroRpc("hiro_offerwall_list", "{}",
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            std::vector<OfferwallItem> offers;
            if (!doc.HasParseError() && doc.IsObject() &&
                doc.HasMember("offers") && doc["offers"].IsArray()) {
                const auto& arr = doc["offers"];
                for (rapidjson::SizeType i = 0; i < arr.Size(); ++i) {
                    const auto& obj = arr[i];
                    if (!obj.IsObject()) continue;
                    OfferwallItem item;
                    item.offerId = obj.HasMember("offer_id") ? obj["offer_id"].GetString() : "";
                    item.title = obj.HasMember("title") ? obj["title"].GetString() : "";
                    item.description = obj.HasMember("description") ? obj["description"].GetString() : "";
                    item.imageUrl = obj.HasMember("image_url") ? obj["image_url"].GetString() : "";
                    item.cost = obj.HasMember("cost") ? obj["cost"].GetInt() : 0;
                    item.currency = obj.HasMember("currency") ? obj["currency"].GetString() : "";
                    item.available = obj.HasMember("available") ? obj["available"].GetBool() : true;
                    offers.push_back(item);
                }
            }
            if (onSuccess) onSuccess(offers);
        }, onError);
}

void IVXHiroSystems::claimOffer(const std::string& offerId,
                                 SuccessCallback onSuccess,
                                 ErrorCallback onError) {
    std::string payload = "{\"offer_id\":\"" + offerId + "\"}";
    callHiroRpc("hiro_offerwall_claim", payload,
        [this, offerId, onSuccess](const std::string&) {
            log("Offer claimed: " + offerId);
            if (onSuccess) onSuccess();
        }, onError);
}

// ---------------------------------------------------------------------------
// Friends
// ---------------------------------------------------------------------------

void IVXHiroSystems::listFriends(int state,
                                  FriendListCallback onSuccess,
                                  ErrorCallback onError) {
    std::string payload = "{\"state\":" + std::to_string(state) + "}";
    callHiroRpc("hiro_friends_list", payload,
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            std::vector<FriendInfo> friends;
            if (!doc.HasParseError() && doc.IsObject() &&
                doc.HasMember("friends") && doc["friends"].IsArray()) {
                const auto& arr = doc["friends"];
                for (rapidjson::SizeType i = 0; i < arr.Size(); ++i) {
                    const auto& obj = arr[i];
                    if (!obj.IsObject()) continue;
                    FriendInfo f;
                    f.userId = obj.HasMember("user_id") ? obj["user_id"].GetString() : "";
                    f.username = obj.HasMember("username") ? obj["username"].GetString() : "";
                    f.displayName = obj.HasMember("display_name") ? obj["display_name"].GetString() : "";
                    f.avatarUrl = obj.HasMember("avatar_url") ? obj["avatar_url"].GetString() : "";
                    f.state = obj.HasMember("state") ? obj["state"].GetInt() : 0;
                    friends.push_back(f);
                }
            }
            if (onSuccess) onSuccess(friends);
        }, onError);
}

void IVXHiroSystems::addFriend(const std::string& userId,
                                SuccessCallback onSuccess,
                                ErrorCallback onError) {
    std::string payload = "{\"user_id\":\"" + userId + "\"}";
    callHiroRpc("hiro_friends_add", payload,
        [this, userId, onSuccess](const std::string&) {
            log("Friend added: " + userId);
            if (onSuccess) onSuccess();
        }, onError);
}

void IVXHiroSystems::removeFriend(const std::string& userId,
                                   SuccessCallback onSuccess,
                                   ErrorCallback onError) {
    std::string payload = "{\"user_id\":\"" + userId + "\"}";
    callHiroRpc("hiro_friends_remove", payload,
        [this, userId, onSuccess](const std::string&) {
            log("Friend removed: " + userId);
            if (onSuccess) onSuccess();
        }, onError);
}

void IVXHiroSystems::blockUser(const std::string& userId,
                                SuccessCallback onSuccess,
                                ErrorCallback onError) {
    std::string payload = "{\"user_id\":\"" + userId + "\"}";
    callHiroRpc("hiro_friends_block", payload,
        [this, userId, onSuccess](const std::string&) {
            log("User blocked: " + userId);
            if (onSuccess) onSuccess();
        }, onError);
}

// ---------------------------------------------------------------------------
// Retention / IAP / Smart ads
// ---------------------------------------------------------------------------

void IVXHiroSystems::getRetention(RetentionCallback onSuccess, ErrorCallback onError) {
    callHiroRpc("hiro_retention_get", "{}",
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            IVXRetentionState state;
            if (!doc.HasParseError() && doc.IsObject()) {
                state.day = doc.HasMember("day") ? doc["day"].GetInt() : 0;
                state.lastLoginAt = doc.HasMember("last_login_at") ? doc["last_login_at"].GetInt64() : 0;
                if (doc.HasMember("rewards") && doc["rewards"].IsArray()) {
                    const auto& arr = doc["rewards"];
                    for (rapidjson::SizeType i = 0; i < arr.Size(); ++i) {
                        const auto& obj = arr[i];
                        if (!obj.IsObject()) continue;
                        IVXRetentionRewardDay r;
                        r.day = obj.HasMember("day") ? obj["day"].GetInt() : 0;
                        r.claimed = obj.HasMember("claimed") ? obj["claimed"].GetBool() : false;
                        state.rewards.push_back(r);
                    }
                }
            }
            if (onSuccess) onSuccess(state);
        }, onError);
}

void IVXHiroSystems::updateRetention(RetentionCallback onSuccess, ErrorCallback onError) {
    callHiroRpc("hiro_retention_update", "{}",
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            IVXRetentionState state;
            if (!doc.HasParseError() && doc.IsObject()) {
                state.day = doc.HasMember("day") ? doc["day"].GetInt() : 0;
                state.lastLoginAt = doc.HasMember("last_login_at") ? doc["last_login_at"].GetInt64() : 0;
                if (doc.HasMember("rewards") && doc["rewards"].IsArray()) {
                    const auto& arr = doc["rewards"];
                    for (rapidjson::SizeType i = 0; i < arr.Size(); ++i) {
                        const auto& obj = arr[i];
                        if (!obj.IsObject()) continue;
                        IVXRetentionRewardDay r;
                        r.day = obj.HasMember("day") ? obj["day"].GetInt() : 0;
                        r.claimed = obj.HasMember("claimed") ? obj["claimed"].GetBool() : false;
                        state.rewards.push_back(r);
                    }
                }
            }
            if (onSuccess) onSuccess(state);
        }, onError);
}

void IVXHiroSystems::checkIapTrigger(const std::string& eventType, IapTriggerCallback onSuccess,
                                    ErrorCallback onError) {
    std::string payload = "{\"event_type\":\"" + eventType + "\"}";
    callHiroRpc("hiro_iap_trigger_check", payload,
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            IVXIapTriggerResult r;
            if (!doc.HasParseError() && doc.IsObject()) {
                r.shouldShow = doc.HasMember("should_show") ? doc["should_show"].GetBool() : false;
                r.offerId = doc.HasMember("offer_id") ? doc["offer_id"].GetString() : "";
                r.discount = doc.HasMember("discount") ? doc["discount"].GetDouble() : 0.0;
                r.expiresAt = doc.HasMember("expires_at") ? doc["expires_at"].GetInt64() : 0;
            }
            if (onSuccess) onSuccess(r);
        }, onError);
}

void IVXHiroSystems::canShowSmartAd(const std::string& placement, SmartAdCallback onSuccess,
                                    ErrorCallback onError) {
    std::string payload = "{\"placement\":\"" + placement + "\"}";
    callHiroRpc("hiro_smart_ad_can_show", payload,
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            IVXSmartAdResult r;
            if (!doc.HasParseError() && doc.IsObject()) {
                r.canShow = doc.HasMember("can_show") ? doc["can_show"].GetBool() : false;
                r.nextAvailableAt = doc.HasMember("next_available_at") ? doc["next_available_at"].GetInt64() : 0;
                r.reason = doc.HasMember("reason") ? doc["reason"].GetString() : "";
            }
            if (onSuccess) onSuccess(r);
        }, onError);
}

// ---------------------------------------------------------------------------
// Internal
// ---------------------------------------------------------------------------

void IVXHiroSystems::callHiroRpc(const std::string& rpcId,
                                  const std::string& payloadJson,
                                  RpcCallback onSuccess,
                                  ErrorCallback onError) {
    IVXManager::getInstance().callRpc(rpcId, payloadJson, onSuccess, onError);
}

void IVXHiroSystems::log(const std::string& message) {
    cocos2d::log("[IntelliVerseX:Hiro] %s", message.c_str());
}

} // namespace IntelliVerseX
